### PR TITLE
Offers obsolete deprecations, update to latest spec

### PR DESCRIPTION
**Pull request by rustyrussell**
I did all the work so that tlvs inside tlvs would "Just work", but then decided I didn't actually want to use that!  (It was for invoices etc inside onion_messages).  It's more precise to demarshal manually so we can give better error feedback.

But there are so many nice generation cleanups here I want to include them.

### CODE REVIEW

File: plugins/keysend.c [hunk 1]
```diff
@@ -360,8 +361,14 @@ static struct command_result *htlc_accepted_call(struct command *cmd,
...
 	}
 
+#if EXPERIMENTAL_FEATURES
+	allowed = cast_const(u64 *, FROMWIRE_TLV_ANY_TYPE);
+#else
 	/* We explicitly allow our type. */
-	allowed[0] = 5482373484;
+	allowed = tal_arr(cmd, u64, 1);
+	allowed[0] = PREIMAGE_TLV_TYPE;
+#endif
```
Comment (cdecker) [inline plugins/keysend.c:370]: Either way the `allowed` array will contain just one entry, and we'll check it below, so we could just add either `FROMWIRE_TLV_ANY_TYPE` or `PREIMAGE_TLV_TYPE` depending on whether we are `EXPERIMENTAL_FEATURES`, saving on allocations and `cast_const` trickeries.

**Review by cdecker — APPROVED** [review]
ACK 78c3cc68fbeeb418057f42c145a97c3db9919f7d

File: plugins/offers_invreq_hook.c [hunk 2]
```diff
@@ -419,7 +419,8 @@ static struct command_result *check_previous_invoice(struct command *cmd,
 /* BOLT-offers #12:
  *  - MUST fail the request if `signature` is not correct.
  */
-static bool check_payer_sig(const struct tlv_invoice_request *invreq,
+static bool check_payer_sig(struct command *cmd,
+			    const struct tlv_invoice_request *invreq,
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply plugins/offers_invreq_hook.c:427]: Oops, some debugging cruft snuck in!

[the remark this answers is not in the document]
Comment (rustyrussell) [reply plugins/offers_invreq_hook.c:427]: Oh, actually, we need cmd here for logging, which the test requires :(

File: plugins/keysend.c (hunk 1 above)
Comment (rustyrussell) [reply plugins/keysend.c:370]: No, FROMWIRE_TLV_ANY_TYPE is literally a magic pointer... I will document this in a comment though.

**Comment by rustyrussell** [comment]
Rebase on master, remove error in last rebase (duplicated code!).

File: lightningd/offer.c [hunk 3]
```diff
@@ -466,16 +467,21 @@ static struct command_result *json_createinvoicerequest(struct command *cmd,
...
 	 */
 	/* This populates the ->fields from our entries */
 	invreq->fields = tlv_make_fields(invreq, invoice_request);
 	merkle_tlv(invreq->fields, &merkle);
-	invreq->payer_signature = tal(invreq, struct bip340sig);
-	hsm_sign_b12(cmd->ld, "invoice_request", "payer_signature",
-		     &merkle, invreq->payer_info, invreq->payer_key,
-		     invreq->payer_signature);
+	invreq->signature = tal(invreq, struct bip340sig);
+	if (deprecated_apis)
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply lightningd/offer.c:477]: This is all experimental, *but* I don't want to break for at least one release.  Yet I want those doing interop testing to get the latest spec version.  (ab)using deprecated_apis here is the least evil solution :(

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/a14c9f3d719d134e56b1f939db51b1335659ebf5

**Pull request merged**